subscriber/common/sink/kafka.go

```diff
@@ -0,0 +1,181 @@
+//  Copyright (c) 2017-2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sink
+
+import (
+	"fmt"
+	"github.com/Shopify/sarama"
+	"github.com/uber-go/tally"
+	"github.com/uber/aresdb/client"
+	"github.com/uber/aresdb/gateway"
+	memCom "github.com/uber/aresdb/memstore/common"
+	"github.com/uber/aresdb/subscriber/common/rules"
+	"github.com/uber/aresdb/subscriber/config"
+	"github.com/uber/aresdb/utils"
+	"go.uber.org/zap"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+)
+
+type KafkaPublisher struct {
+	sarama.SyncProducer
+	client.UpsertBatchBuilder
+
+	ServiceConfig config.ServiceConfig
+	JobConfig     *rules.JobConfig
+	Scope         tally.Scope
+	ClusterName   string
+}
+
+func NewKafkaPublisher(serviceConfig config.ServiceConfig, jobConfig *rules.JobConfig, cluster string, sinkCfg config.SinkConfig) (Sink, error) {
+	if sinkCfg.GetSinkMode() != config.Sink_Kafka {
+		return nil, fmt.Errorf("Failed to NewKafkaPublisher, wrong sinkMode=%d", sinkCfg.GetSinkMode())
+	}
+
+	addresses := strings.Split(sinkCfg.KafkaProducerConfig.Brokers, ",")
+	serviceConfig.Logger.Info("Kafka borkers address", zap.Any("brokers", addresses))
+
+	cfg := sarama.NewConfig()
+	cfg.Producer.RequiredAcks = sarama.WaitForAll // Wait for all in-sync replicas to ack the message
+	if sinkCfg.KafkaProducerConfig.RetryMax > 0 {
+		cfg.Producer.Retry.Max = sinkCfg.KafkaProducerConfig.RetryMax
+	}
+	if sinkCfg.KafkaProducerConfig.TimeoutInSec > 0 {
+		cfg.Producer.Timeout = time.Second * time.Duration(sinkCfg.KafkaProducerConfig.TimeoutInSec)
+	}
+	cfg.Producer.Return.Successes = true
+
+	p, err := sarama.NewSyncProducer(addresses, cfg)
+	if err != nil {
+		return nil, utils.StackError(err, "Unable to initialize Kafka producer")
+	}
+
+	aresControllerClient := gateway.NewControllerHTTPClient(serviceConfig.ControllerConfig.Address,
+		time.Duration(serviceConfig.ControllerConfig.Timeout)*time.Second,
+		http.Header{
+			"RPC-Caller":  []string{os.Getenv("UDEPLOY_APP_ID")},
```

Comment:
this is internal

Reply:
the aresControllerClient can be intialized from config file, and injected here instead. that way you can remove uber  specific headers from here

Reply:
it will cause import cycle if aresControllerClient is intialized from config file. 
Anyway, I make it internal by changing lot of apis